src/persist/src/indexed/encoding.rs

```diff
@@ -0,0 +1,469 @@
+                if prev.upper() != desc.lower() {
+                    return Err(format!(
+                        "invalid batch sequence: {:?} followed by {:?}",
+                        prev, desc
+                    )
+                    .into());
+                }
+            }
+            prev = Some(desc)
+        }
+        Ok(())
+    }
+}
+
+impl BlobTraceMeta {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<&Description<u64>> = None;
+        for (desc, _) in self.batches.iter() {
+            if let Some(prev) = prev {
+                if prev.upper() != desc.lower() {
+                    return Err(format!(
+                        "invalid batch sequence: {:?} followed by {:?}",
+                        prev, desc,
+                    )
+                    .into());
+                }
+            }
+            prev = Some(desc)
+        }
+        Ok(())
+    }
+}
+
+impl BlobFutureBatch {
+    /// Asserts Self's documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<(&u64, &String, &String)> = None;
+        for (seqno, (key, val), ts, _) in self.updates.iter() {
+            // Check seqno against desc.
+            if !self.desc.lower().less_equal(seqno) {
+                return Err(format!(
+                    "seqno {:?} is before the batch lower: {:?}",
+                    seqno, self.desc
+                )
+                .into());
+            }
+            if self.desc.upper().less_equal(seqno) {
+                return Err(format!(
+                    "seqno {:?} is after or equal to the batch upper: {:?}",
+                    seqno, self.desc
+                )
+                .into());
+            }
+
+            // Check ordering.
+            let this = (ts, key, val);
+            if let Some(prev) = prev {
+                match prev.cmp(&this) {
+                    Ordering::Less => {} // Correct.
+                    Ordering::Equal => return Err(format!("unconsolidated: {:?}", this).into()),
+                    Ordering::Greater => {
+                        return Err(format!("unsorted: {:?} was before {:?}", prev, this).into())
+                    }
+                }
+            }
+            prev = Some(this);
+        }
+
+        Ok(())
+    }
+}
+
+impl BlobTraceBatch {
+    /// Asserts the documented invariants, returning an error if any are
+    /// violated.
+    pub fn validate(&self) -> Result<(), Error> {
+        let mut prev: Option<(&String, &String, &u64)> = None;
+        for ((key, val), ts, _) in self.updates.iter() {
+            // Check ts against desc.
+            if !self.desc.lower().less_equal(ts) {
+                return Err(format!(
+                    "timestamp {} is in advance of the batch lower: {:?}",
```

Comment:
should this be "timestamp {} is not at or in advance of the batch lower" because if we hit this branch it means that no element in `desc` is less than or equal to `ts` (I read "in advance of" as "after / beyond / greater than")

Reply:
I found out yesterday that my understanding of what "in advance of" means is literally flipped. Changed this to less/greater/equal since I'm not convinced the additional correctness of the advance terminology is worth the tradeoffs of common understandability

Reply:
I found out yesterday that my understanding of "in advance of" is actually backward! Changed this to less/greater

Reply:
I don't think your understanding is wrong though -- like "I filled out my taxes in advance of the deadline and wasn't late!" is a legitimate and sensical sentence. Unfortunately theres a different, opposite idiom for frontiers :D